File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/push/exceptions/NonSuccessfulResponseCodeException.java
/**
 * Copyright (C) 2014-2016 Open Whisper Systems
 *
 * Licensed according to the LICENSE file in this repository.
 */

package org.whispersystems.signalservice.api.push.exceptions;

import java.io.IOException;

/**
 * Indicates a server response that is not successful, typically something outside the 2xx range.
 */
public class NonSuccessfulResponseCodeException extends IOException {

  private final int    code;
  private final String body;

  public NonSuccessfulResponseCodeException(int code) {
    super("StatusCode: " + code);
    this.code = code;
    this.body = null;
  }

  public NonSuccessfulResponseCodeException(int code, String s) {
    super("[" + code + "] " + s);
    this.code = code;
    this.body = null;
  }

  public NonSuccessfulResponseCodeException(int code, String s, String body) {
    super("[" + code + "] " + s);
    this.code = code;
    this.body = body;
  }

  public int getCode() {
    return code;
  }

  public boolean is4xx() {
    return code >= 400 && code < 500;
  }

  public boolean is5xx() {
    return code >= 500 && code < 600;
  }

  public String getBody() {
    return body;
  }
}

File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/push/exceptions/NonSuccessfulResponseCodeException.kt
/**
 * Copyright (C) 2014-2016 Open Whisper Systems
 *
 * Licensed according to the LICENSE file in this repository.
 */
package org.whispersystems.signalservice.api.push.exceptions

import java.io.IOException

/**
 * Indicates a server response that is not successful, typically something outside the 2xx range.
 */
open class NonSuccessfulResponseCodeException : IOException {
  @JvmField
  val code: Int
  val stringBody: String?
  val binaryBody: ByteArray?
  val headers: Map<String, String>

  constructor(code: Int) : super("StatusCode: $code") {
    this.code = code
    this.stringBody = null
    this.binaryBody = null
    this.headers = emptyMap()
  }

  constructor(code: Int, message: String) : super("[$code] $message") {
    this.code = code
    this.stringBody = null
    this.binaryBody = null
    this.headers = emptyMap()
  }

  @JvmOverloads
  constructor(code: Int, message: String, body: String?, headers: Map<String, String> = emptyMap()) : super("[$code] $message") {
    this.code = code
    this.stringBody = body
    this.binaryBody = null
    this.headers = headers.mapKeys { it.key.lowercase() }
  }

  @JvmOverloads
  constructor(code: Int, message: String, body: ByteArray?, headers: Map<String, String> = emptyMap()) : super("[$code] $message") {
    this.code = code
    this.stringBody = null
    this.binaryBody = body
    this.headers = headers.mapKeys { it.key.lowercase() }
  }

  fun is4xx(): Boolean {
    return code >= 400 && code < 500
  }

  fun is5xx(): Boolean {
    return code >= 500 && code < 600
  }
}
